File: tests/test_transient_reconnect.py
"""T-33 5–20s network-loss reconnect — focused contract and node tests."""

from __future__ import annotations

import shutil
import subprocess
from datetime import timedelta
from pathlib import Path

import pytest
from asgiref.sync import async_to_sync
from channels.layers import channel_layers
from django.core.cache import cache
from django.utils import timezone

from apps.events.call_room_logic import (
    RECONNECT_INITIAL_DELAY_MS,
    RECONNECT_MAX_DELAY_MS,
    RECONNECT_WINDOW_MS,
    CallRoomPhase,
    CallRoomState,
    compute_remaining_ms,
    next_reconnect_delay_ms,
    reconnect_window_exhausted,
)
from apps.events.models import Round
from apps.events.orchestrator import OrchestratorRealtime
from apps.events.scheduler import RoundScheduler
from apps.protocol.constants import MessageType
from tests.test_orchestrator_realtime import _make_event, _pair, _participant
from tests.test_websocket_auth import _connect, _cookie_for, _hello

REPO = Path(__file__).resolve().parents[1]
NEGOTIATOR_JS = (
    REPO / "apps" / "events" / "static" / "events" / "js" / "perfect_negotiator.js"
)
NEGOTIATOR_JS_COPY = REPO / "apps" / "events" / "js" / "perfect_negotiator.js"
CALL_ROOM_JS = REPO / "apps" / "events" / "static" / "events" / "js" / "call_room.js"
VIDEO_ROOM_HTML = REPO / "apps" / "events" / "templates" / "events" / "video_room.html"
NODE_TEST = REPO / "tests" / "js" / "test_transient_reconnect.mjs"


def _pairing_payload(**overrides):
    base = {
        "round_number": 1,
        "room_id": "room-1",
        "partner_id": "00000000-0000-0000-0000-000000000002",
        "is_offerer": True,
        "round_start_ts": 1_700_000_000_000,
        "round_end_ts": 1_700_000_300_000,
        "partner_display_name": "Sara",
        "partner_tags": ["ai"],
    }
    base.update(overrides)
    return base


def test_negotiator_copies_stay_in_sync():
    assert NEGOTIATOR_JS.read_text(encoding="utf-8") == NEGOTIATOR_JS_COPY.read_text(
        encoding="utf-8"
    )


def test_backoff_is_bounded_exponential():
    assert next_reconnect_delay_ms(0) == 500
    assert next_reconnect_delay_ms(1) == 1000
    assert next_reconnect_delay_ms(2) == 2000
    assert next_reconnect_delay_ms(3) == 4000
    assert next_reconnect_delay_ms(8) == 4000
    assert RECONNECT_INITIAL_DELAY_MS == 500
    assert RECONNECT_MAX_DELAY_MS == 4000
    assert RECONNECT_WINDOW_MS == 45_000
    assert reconnect_window_exhausted(0, 44_999) is False
    assert reconnect_window_exhausted(0, 45_000) is True


def test_call_room_reconnect_contract_reuses_t14_and_t31():
    src = CALL_ROOM_JS.read_text(encoding="utf-8")
    assert "client.hello" in src
    assert "/ws/events/" in src
    assert "nextReconnectDelayMs" in src
    assert "RECONNECTING" in src
    assert "recoverAfterSignalingRestore" in src
    assert "navigator.mediaDevices.getUserMedia" not in src
    assert "_giveUpReconnect" in src
    assert "online" in src
    assert src.count("new WebSocket") <= 1

    negotiator = NEGOTIATOR_JS.read_text(encoding="utf-8")
    assert "recoverAfterSignalingRestore" in negotiator
    assert "_maybeRestartIce" in negotiator
    assert (
        "getUserMedia"
        not in negotiator.split("recoverAfterSignalingRestore()")[1][:400]
    )


def test_video_room_exposes_reconnecting_state():
    html = VIDEO_ROOM_HTML.read_text(encoding="utf-8")
    assert "reconnecting" in html
    assert 'id="connectionBadge"' in html
    assert 'id="timerDisplay"' in html
    assert "statusFooter" in html


def test_transient_disconnect_preserves_partner_round_and_timer():
    state = CallRoomState().on_pairing(_pairing_payload())
    state = state.on_round_start({"round_number": 1})
    disconnected = state.on_transient_disconnect()
    assert disconnected.phase == CallRoomPhase.IN_ROUND
    assert disconnected.phase != CallRoomPhase.EVENT_ENDED
    assert disconnected.round_number == 1
    assert disconnected.round_end_ts == 1_700_000_300_000
    assert disconnected.partner is not None
    assert disconnected.partner.room_id == "room-1"

    remaining = compute_remaining_ms(disconnected.round_end_ts, 0, 1_700_000_100_000)
    remaining_again = compute_remaining_ms(
        disconnected.round_end_ts, 0, 1_700_000_100_000
    )
    assert remaining == remaining_again == 200_000


def test_idle_disconnect_still_uses_disconnected_phase():
    state = CallRoomState().on_transient_disconnect()
    assert state.phase == CallRoomPhase.DISCONNECTED
    assert state.phase != CallRoomPhase.EVENT_ENDED


def test_reconnect_snapshot_is_isolated_to_one_participant(setup_round):
    event = setup_round["event"]
    p1 = setup_round["p1"]
    p3 = setup_round["p3"]
    r1 = setup_round["r1"]
    now = r1.starts_at + timedelta(seconds=1)
    orch = OrchestratorRealtime(event, now_fn=lambda: now)

    messages = orch.reconnect_snapshot_messages(p1, now=now)
    types = [msg["type"] for msg in messages]
    assert MessageType.SERVER_PAIRING in types
    assert MessageType.SERVER_ROUND_START in types
    assert MessageType.SERVER_EVENT_END not in types
    assert MessageType.SERVER_ROUND_END not in types
    pairing = next(msg for msg in messages if msg["type"] == MessageType.SERVER_PAIRING)
    assert pairing["payload"]["partner_id"] == str(setup_round["p2"].pk)
    assert pairing["payload"]["room_id"] == "room-r1-a"
    assert pairing["payload"]["round_number"] == 1
    assert pairing["payload"]["round_end_ts"] == int(r1.ends_at.timestamp() * 1000)

    other_room = orch.reconnect_snapshot_messages(p3, now=now)
    other_pairing = next(
        msg for msg in other_room if msg["type"] == MessageType.SERVER_PAIRING
    )
    assert other_pairing["payload"]["room_id"] == "room-r1-b"
    assert other_pairing["payload"]["partner_id"] == str(setup_round["p4"].pk)


def test_reconnect_snapshot_does_not_end_a_live_round(setup_round):
    event = setup_round["event"]
    r1 = setup_round["r1"]
    now = r1.starts_at + timedelta(seconds=5)
    orch = OrchestratorRealtime(event, now_fn=lambda: now)
    messages = orch.reconnect_snapshot_messages(setup_round["p1"], now=now)
    assert all(msg["type"] != MessageType.SERVER_EVENT_END for msg in messages)
    assert all(msg["type"] != MessageType.SERVER_ROUND_END for msg in messages)


def test_reconnect_after_event_completion_replays_event_end(setup_round):
    event = setup_round["event"]
    last = setup_round["r2"]
    now = last.ends_at + timedelta(seconds=1)
    orch = OrchestratorRealtime(event, now_fn=lambda: now)
    messages = orch.reconnect_snapshot_messages(setup_round["p1"], now=now)
    assert [msg["type"] for msg in messages] == [MessageType.SERVER_EVENT_END]


@pytest.mark.django_db(transaction=True)
def test_t14_reconnect_replays_snapshot_only_on_that_socket(setup_round, settings):
    settings.CHANNEL_LAYERS = {
        "default": {"BACKEND": "channels.layers.InMemoryChannelLayer"}
    }
    channel_layers.backends = {}
    p1 = setup_round["p1"]
    p3 = setup_round["p3"]
    cache.clear()
    now = timezone.now()
    r1 = setup_round["r1"]
    r2 = setup_round["r2"]
    r1.starts_at = now - timedelta(seconds=5)
    r1.ends_at = now + timedelta(minutes=5)
    r1.save(update_fields=["starts_at", "ends_at"])
    r2.starts_at = now + timedelta(minutes=10)
    r2.ends_at = now + timedelta(minutes=15)
    r2.save(update_fields=["starts_at", "ends_at"])
    cookie_a = _cookie_for(p1)
    cookie_c = _cookie_for(p3)

    async def scenario():
        first = _connect(cookie_a)
        other = _connect(cookie_c)
        assert (await first.connect())[0]
        assert (await other.connect())[0]
        await first.send_json_to(_hello())
        hello = await first.receive_json_from()
        assert hello["type"] == MessageType.SERVER_HELLO
        await other.send_json_to(_hello())
        assert (await other.receive_json_from())["type"] == MessageType.SERVER_HELLO
        await first.disconnect()

        second = _connect(cookie_a)
        assert (await second.connect())[0]
        await second.send_json_to(_hello())
        hello2 = await second.receive_json_from()
        assert hello2["type"] == MessageType.SERVER_HELLO
        pairing = await second.receive_json_from()
        start = await second.receive_json_from()
        assert pairing["type"] == MessageType.SERVER_PAIRING
        assert pairing["payload"]["room_id"] == "room-r1-a"
        assert pairing["payload"]["partner_id"] == str(setup_round["p2"].pk)
        assert start["type"] == MessageType.SERVER_ROUND_START
        assert start["payload"]["room_id"] == "room-r1-a"

        await other.send_json_to(
            {
                "type": "client.ping",
                "version": 1,
                "payload": {"client_ts": 1},
            }
        )
        pong = await other.receive_json_from()
        assert pong["type"] == MessageType.SERVER_PONG
        await second.disconnect()
        await other.disconnect()

    async_to_sync(scenario)()


@pytest.fixture
def setup_round(db):
    event = _make_event()
    p1 = _participant(event, "Alice")
    p2 = _participant(event, "Bob")
    p3 = _participant(event, "Cara")
    p4 = _participant(event, "Drew")
    scheduler = RoundScheduler(event)
    scheduler.persist_schedule()
    r1 = Round.objects.get(event=event, number=1)
    r2 = Round.objects.get(event=event, number=2)
    pair_1 = _pair(event, r1, p1, p2, "room-r1-a")
    pair_2 = _pair(event, r1, p3, p4, "room-r1-b")
    return {
        "event": event,
        "p1": p1,
        "p2": p2,
        "p3": p3,
        "p4": p4,
        "r1": r1,
        "r2": r2,
        "pair_1": pair_1,
        "pair_2": pair_2,
    }


@pytest.mark.skipif(
    shutil.which("node") is None, reason="node is required for JS reconnect tests"
)
def test_node_transient_reconnect():
    completed = subprocess.run(
        ["node", "--test", str(NODE_TEST)],
        cwd=REPO,
        capture_output=True,
        text=True,
        timeout=30,
        check=False,
    )
    assert completed.returncode == 0, completed.stdout + completed.stderr

File: tests/js/test_transient_reconnect.mjs
/**
 * T-33: 5–20s network loss reconnects the T-14 socket with bounded backoff
 * to the same partner/room/round without resetting the timer or media.
 */
import assert from "node:assert/strict";
import path from "node:path";
import { afterEach, before, beforeEach, test } from "node:test";
import { fileURLToPath, pathToFileURL } from "node:url";

class FakeTrack {
  constructor(kind) {
    this.kind = kind;
    this.id = `${kind}-${Math.random().toString(36).slice(2, 8)}`;
    this.enabled = true;
    this.readyState = "live";
  }
  stop() {
    this.readyState = "ended";
  }
  getSettings() {
    return { width: 640, height: 360 };
  }
}

class FakeStream {
  constructor(tracks = []) {
    this._tracks = [...tracks];
    this.id = `stream-${Math.random().toString(36).slice(2, 8)}`;
  }
  getTracks() {
    return [...this._tracks];
  }
  getAudioTracks() {
    return this._tracks.filter((track) => track.kind === "audio");
  }
  getVideoTracks() {
    return this._tracks.filter((track) => track.kind === "video");
  }
  addTrack(track) {
    this._tracks.push(track);
  }
}

class FakePC {
  constructor(config = {}) {
    this.config = config;
    this.signalingState = "stable";
    this.connectionState = "new";
    this._senders = [];
    this.onnegotiationneeded = null;
    this.onicecandidate = null;
    this.ontrack = null;
    this.onconnectionstatechange = null;
    this.localDescription = null;
    this.remoteDescription = null;
    this.addTrackCalls = 0;
    this.restartIceCalls = 0;
    this.lastOfferOptions = null;
    this.closed = false;
    FakePC.instances.push(this);
  }
  addTrack(track, stream) {
    this.addTrackCalls += 1;
    const sender = {
      track,
      stream,
      _trackKind: track.kind,
      replaceTrack: async (next) => {
        sender.track = next;
      },
      getParameters: () => ({ encodings: [{}] }),
      setParameters: async () => {},
    };
    this._senders.push(sender);
    queueMicrotask(() => this.onnegotiationneeded?.());
    return sender;
  }
  getSenders() {
    return this._senders;
  }
  removeTrack(sender) {
    this._senders = this._senders.filter((item) => item !== sender);
  }
  restartIce() {
    this.restartIceCalls += 1;
    queueMicrotask(() => this.onnegotiationneeded?.());
  }
  close() {
    this.closed = true;
    this.connectionState = "closed";
    this.signalingState = "closed";
    this.onconnectionstatechange?.();
  }
  async setLocalDescription(desc) {
    this.localDescription = desc || { type: "offer", sdp: "fake-sdp" };
    if (!desc || desc.type === "offer") this.signalingState = "have-local-offer";
    if (desc?.type === "answer" || desc?.type === "rollback") {
      this.signalingState = "stable";
    }
  }
  async createOffer(options) {
    this.lastOfferOptions = options || null;
    return {
      type: "offer",
      sdp: options?.iceRestart ? "fake-offer-ice-restart" : "fake-offer",
    };
  }
  async setRemoteDescription(desc) {
    this.remoteDescription = desc;
    this.signalingState = desc.type === "offer" ? "have-remote-offer" : "stable";
  }
  async addIceCandidate() {}
  async getStats() {
    return new Map();
  }
  getTransceivers() {
    return [];
  }
}
FakePC.instances = [];

class FakeWebSocket {
  static CONNECTING = 0;
  static OPEN = 1;
  static CLOSING = 2;
  static CLOSED = 3;
  constructor(url) {
    this.url = url;
    this.readyState = FakeWebSocket.CONNECTING;
    this.sent = [];
    this.onopen = null;
    this.onmessage = null;
    this.onclose = null;
    this.onerror = null;
    FakeWebSocket.instances.push(this);
  }
  send(data) {
    this.sent.push(typeof data === "string" ? JSON.parse(data) : data);
  }
  close() {
    if (this.readyState === FakeWebSocket.CLOSED) return;
    this.readyState = FakeWebSocket.CLOSED;
    this.onclose?.();
  }
  open() {
    this.readyState = FakeWebSocket.OPEN;
    this.onopen?.();
  }
}
FakeWebSocket.instances = [];

function installBrowserMocks() {
  globalThis.MediaStream = FakeStream;
  globalThis.RTCPeerConnection = FakePC;
  globalThis.WebSocket = FakeWebSocket;
  globalThis.window = globalThis;
  globalThis.location = { protocol: "http:", host: "localhost" };

  const listeners = new Map();
  globalThis.addEventListener = (type, handler) => {
    const list = listeners.get(type) || [];
    list.push(handler);
    listeners.set(type, list);
  };
  globalThis.removeEventListener = (type, handler) => {
    const list = listeners.get(type) || [];
    listeners.set(
      type,
      list.filter((item) => item !== handler),
    );
  };
  globalThis.__dispatch = (type) => {
    (listeners.get(type) || []).forEach((handler) => handler());
  };
  globalThis.__listeners = listeners;

  const getUserMedia = async () => {
    globalThis.__gumCalls = (globalThis.__gumCalls || 0) + 1;
    return new FakeStream([new FakeTrack("audio"), new FakeTrack("video")]);
  };

  const nav = globalThis.navigator;
  if (nav?.mediaDevices) {
    nav.mediaDevices.getUserMedia = getUserMedia;
  } else if (nav) {
    Object.defineProperty(nav, "mediaDevices", {
      configurable: true,
      value: { getUserMedia },
    });
  } else {
    Object.defineProperty(globalThis, "navigator", {
      configurable: true,
      value: { mediaDevices: { getUserMedia } },
    });
  }
}

installBrowserMocks();

const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const negotiatorHref = pathToFileURL(
  path.join(repoRoot, "apps/events/static/events/js/perfect_negotiator.js"),
).href;
const callRoomHref = pathToFileURL(
  path.join(repoRoot, "apps/events/static/events/js/call_room.js"),
).href;

const negotiatorMod = await import(negotiatorHref);
const callRoomMod = await import(callRoomHref);

const {
  PerfectNegotiator,
  resetSharedLocalMediaForTests,
  getSharedGetUserMediaCallCount,
} = negotiatorMod;
const {
  CallRoomController,
  CallRoomPhase,
  computeRemainingMs,
  nextReconnectDelayMs,
  reconnectWindowExhausted,
  RECONNECT_INITIAL_DELAY_MS,
  RECONNECT_MAX_DELAY_MS,
  RECONNECT_WINDOW_MS,
} = callRoomMod;

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function pairingPayload(roundNumber, roomId, partnerId, roundEndTs) {
  return {
    round_number: roundNumber,
    room_id: roomId,
    partner_id: partnerId,
    is_offerer: true,
    round_start_ts: Date.now(),
    round_end_ts: roundEndTs ?? Date.now() + 120_000,
    partner_display_name: `Partner ${roundNumber}`,
    partner_tags: ["ai"],
  };
}

function setConn(pc, state) {
  pc.connectionState = state;
  pc.onconnectionstatechange?.();
}

function uiElements() {
  return {
    connectionBadge: { textContent: "", className: "", dataset: {} },
    qualityBanner: { hidden: true, textContent: "" },
    timerDisplay: { textContent: "", dataset: {} },
    timerContainer: { dataset: {} },
    errorBanner: { hidden: true, textContent: "" },
    localVideo: { srcObject: null },
    remoteVideo: { srcObject: null },
    partnerName: { textContent: "" },
    roundLabel: { textContent: "" },
    roomLabel: { textContent: "" },
    phaseBadge: { textContent: "", dataset: {} },
    statusFooter: { textContent: "" },
  };
}

const liveControllers = [];

function makeController(overrides = {}) {
  const elements = overrides.elements || uiElements();
  const controller = new CallRoomController({
    myParticipantId: "participant-a",
    reconnectInitialDelayMs: 12,
    reconnectMaxDelayMs: 40,
    reconnectWindowMs: 120,
    createWebSocket: (url) => new FakeWebSocket(url),
    elements,
    ...overrides,
  });
  liveControllers.push(controller);
  return controller;
}

async function enterRound(controller, payload = pairingPayload(1, "room-1", "partner-b")) {
  controller.clockSync.offsetMs = 0;
  controller.clockSync.status = "synchronised";
  await controller.ensureLocalMedia();
  await controller._onPairing(payload);
  await controller._onRoundStart({ round_number: payload.round_number });
  return payload;
}

before(() => {
  FakePC.instances = [];
  FakeWebSocket.instances = [];
});

beforeEach(() => {
  FakePC.instances = [];
  FakeWebSocket.instances = [];
  globalThis.__gumCalls = 0;
  globalThis.__listeners?.clear();
  resetSharedLocalMediaForTests();
});

afterEach(() => {
  for (const controller of liveControllers) {
    try {
      controller.disconnect();
    } catch {
      // ignore
    }
  }
  liveControllers.length = 0;
  resetSharedLocalMediaForTests();
});

test("bounded exponential backoff caps at max delay", () => {
  assert.equal(nextReconnectDelayMs(0, 500, 4000), 500);
  assert.equal(nextReconnectDelayMs(1, 500, 4000), 1000);
  assert.equal(nextReconnectDelayMs(2, 500, 4000), 2000);
  assert.equal(nextReconnectDelayMs(3, 500, 4000), 4000);
  assert.equal(nextReconnectDelayMs(6, 500, 4000), 4000);
  assert.equal(RECONNECT_INITIAL_DELAY_MS, 500);
  assert.equal(RECONNECT_MAX_DELAY_MS, 4000);
  assert.equal(RECONNECT_WINDOW_MS, 45_000);
  assert.equal(reconnectWindowExhausted(0, 44_999, 45_000), false);
  assert.equal(reconnectWindowExhausted(0, 45_000, 45_000), true);
});

test("signaling drop shows RECONNECTING and keeps the same partner/round/timer", async () => {
  const elements = uiElements();
  const controller = makeController({ elements });
  const payload = await enterRound(controller);
  const pc = controller.negotiator.pc;
  const stream = controller.negotiator.localStream;
  const roundEndTs = controller.roundEndTs;
  setConn(pc, "connected");

  const remainingBefore = computeRemainingMs(roundEndTs, 0, Date.now());
  controller._onSignalingClosed();

  assert.equal(controller.phase, CallRoomPhase.IN_ROUND);
  assert.equal(controller.roundNumber, 1);
  assert.equal(controller.partner.partnerId, "partner-b");
  assert.equal(controller.partner.roomId, "room-1");
  assert.equal(controller.roundEndTs, roundEndTs);
  assert.ok(controller._timerInterval);
  assert.equal(elements.connectionBadge.textContent, "RECONNECTING");
  assert.equal(elements.connectionBadge.dataset.quality, "reconnecting");
  assert.match(elements.qualityBanner.textContent, /RECONNECTING/);
  assert.equal(controller.negotiator.pc, pc);
  assert.equal(controller.negotiator.localStream, stream);
  assert.equal(getSharedGetUserMediaCallCount(), 1);

  await delay(30);
  const remainingDuring = computeRemainingMs(controller.roundEndTs, 0, Date.now());
  assert.ok(remainingDuring <= remainingBefore);
  assert.ok(remainingDuring > remainingBefore - 2000);
  controller.disconnect();
});

test("backoff retries T-14 hello on a new socket and ignores the stale one", async () => {
  const controller = makeController();
  controller.connect();
  assert.equal(FakeWebSocket.instances.length, 1);
  const first = FakeWebSocket.instances[0];
  first.open();
  assert.equal(first.sent[0].type, "client.hello");

  first.close();
  assert.equal(controller.phase, CallRoomPhase.DISCONNECTED);
  await delay(20);

  assert.ok(FakeWebSocket.instances.length >= 2);
  const second = FakeWebSocket.instances[1];
  assert.equal(first.onclose, null);
  assert.equal(first.onmessage, null);
  assert.notEqual(second, first);

  const generation = controller._wsGeneration;
  first.onclose?.();
  first.close();
  assert.equal(controller._wsGeneration, generation);

  second.open();
  assert.equal(second.sent[0].type, "client.hello");
  assert.equal(second.sent[0].version, 1);
  controller.disconnect();
});

test("hello after in-round drop restores the same PC/media and kicks T-32 ICE restart", async () => {
  const elements = uiElements();
  const controller = makeController({ elements });
  await enterRound(controller);
  const pc = controller.negotiator.pc;
  const stream = controller.negotiator.localStream;
  const roundEndTs = controller.roundEndTs;
  const pcCount = FakePC.instances.length;
  setConn(pc, "connected");
  setConn(pc, "failed");
  controller._onSignalingClosed();

  assert.equal(elements.connectionBadge.textContent, "RECONNECTING");
  assert.equal(controller.phase, CallRoomPhase.IN_ROUND);

  await delay(30);
  controller._onServerHello({ server_ts: Date.now() });

  assert.equal(controller.phase, CallRoomPhase.IN_ROUND);
  assert.equal(controller.roundNumber, 1);
  assert.equal(controller.roundEndTs, roundEndTs);
  assert.equal(controller.partner.partnerId, "partner-b");
  assert.equal(controller.negotiator.pc, pc);
  assert.equal(controller.negotiator.localStream, stream);
  assert.equal(FakePC.instances.length, pcCount);
  assert.ok(pc.restartIceCalls >= 1);
  assert.equal(getSharedGetUserMediaCallCount(), 1);
  assert.equal(globalThis.__gumCalls, 1);
  assert.ok(stream.getTracks().every((track) => track.readyState === "live"));
  controller.disconnect();
});

test("same pairing snapshot does not create a second PeerConnection or offer", async () => {
  const controller = makeController();
  const payload = await enterRound(controller);
  const pc = controller.negotiator.pc;
  const pcCount = FakePC.instances.length;
  setConn(pc, "connected");
  controller._onSignalingClosed();

  await controller._onPairing(payload);
  await controller._onRoundStart({ round_number: 1 });

  assert.equal(controller.phase, CallRoomPhase.IN_ROUND);
  assert.equal(controller.negotiator.pc, pc);
  assert.equal(FakePC.instances.length, pcCount);
  assert.equal(getSharedGetUserMediaCallCount(), 1);
  controller.disconnect();
});

test("retry window closes the stale socket without ending the event", async () => {
  const elements = uiElements();
  const controller = makeController({
    elements,
    reconnectInitialDelayMs: 5,
    reconnectMaxDelayMs: 5,
    reconnectWindowMs: 20,
  });
  await enterRound(controller);
  const roundEndTs = controller.roundEndTs;
  controller._onSignalingClosed();
  await delay(50);

  assert.equal(controller._reconnectGaveUp, true);
  assert.equal(controller.phase, CallRoomPhase.IN_ROUND);
  assert.notEqual(controller.phase, CallRoomPhase.EVENT_ENDED);
  assert.equal(controller.roundEndTs, roundEndTs);
  assert.equal(controller.ws, null);
  assert.equal(elements.connectionBadge.textContent, "FAILED");
  assert.equal(elements.errorBanner.hidden, false);
  assert.ok(controller._timerInterval);
  controller.disconnect();
});

test("online event retries immediately after the window was exhausted", async () => {
  const controller = makeController({
    reconnectInitialDelayMs: 5,
    reconnectMaxDelayMs: 5,
    reconnectWindowMs: 15,
  });
  await enterRound(controller);
  controller.connect();
  FakeWebSocket.instances[0].open();
  controller._onSignalingClosed();
  await delay(40);
  assert.equal(controller._reconnectGaveUp, true);
  const before = FakeWebSocket.instances.length;

  globalThis.__dispatch("online");
  assert.equal(controller._reconnectGaveUp, false);
  assert.ok(FakeWebSocket.instances.length > before);
  controller.disconnect();
});

test("one participant reconnecting does not pause another room's timer", async () => {
  const roomA = makeController({ elements: uiElements() });
  const roomB = makeController({
    myParticipantId: "participant-c",
    elements: uiElements(),
  });
  await enterRound(roomA, pairingPayload(1, "room-1", "partner-b", Date.now() + 90_000));
  await enterRound(roomB, pairingPayload(1, "room-2", "partner-d", Date.now() + 90_000));

  const endA = roomA.roundEndTs;
  const endB = roomB.roundEndTs;
  const remainingB = computeRemainingMs(endB, 0, Date.now());
  roomA._onSignalingClosed();

  assert.equal(roomA.phase, CallRoomPhase.IN_ROUND);
  assert.equal(roomB.phase, CallRoomPhase.IN_ROUND);
  assert.equal(roomB.roundEndTs, endB);
  assert.equal(roomA.roundEndTs, endA);
  assert.equal(roomB.elements.connectionBadge.textContent, "");
  const remainingBAfter = computeRemainingMs(roomB.roundEndTs, 0, Date.now());
  assert.ok(remainingBAfter <= remainingB);
  assert.ok(remainingBAfter > remainingB - 2000);
  roomA.disconnect();
  roomB.disconnect();
});

test("event_end still tears down and does not reconnect", async () => {
  const controller = makeController();
  controller.connect();
  FakeWebSocket.instances[0].open();
  await enterRound(controller);
  await controller._onEventEnd({ reason: "completed" });
  const sockets = FakeWebSocket.instances.length;
  await delay(30);
  assert.equal(controller.phase, CallRoomPhase.EVENT_ENDED);
  assert.equal(FakeWebSocket.instances.length, sockets);
  assert.equal(controller.ws, null);
  assert.equal(controller._reconnectTimer, null);
});

test("recoverAfterSignalingRestore reuses T-32 ICE restart without getUserMedia", async () => {
  const negotiator = new PerfectNegotiator({
    myParticipantId: "participant-a",
    partnerParticipantId: "participant-b",
    roomId: "room-1",
    sendSignalingMessage: () => {},
  });
  await negotiator.acquireLocalMedia();
  const pc = negotiator.pc;
  const stream = negotiator.localStream;
  setConn(pc, "failed");
  negotiator.recoverAfterSignalingRestore();
  assert.equal(negotiator.pc, pc);
  assert.ok(pc.restartIceCalls >= 1);
  assert.equal(negotiator.localStream, stream);
  assert.equal(getSharedGetUserMediaCallCount(), 1);
  negotiator.leave();
});
